Source/modules/webcl/WebCL.cpp

```diff
@@ -0,0 +1,376 @@
+
+    if (!clDeviceList.size()) {
+        es.throwWebCLException(WebCLException::INVALID_DEVICE, WebCLException::invalidDeviceMessage);
+        return nullptr;
+    }
+
+    clContextId = clCreateContext(nullptr, clDeviceList.size(), clDeviceList.data(), nullptr, nullptr, &err);
+    if (err != CL_SUCCESS) {
+        WebCLException::throwException(err, es);
+        return nullptr;
+    }
+
+    // Check all the enabled extensions and cache it to avoid enabling after context creation.
+    HashSet<String> enabledExtensions;
+    getAllEnabledExtensions(this, devices[0]->getPlatform(), devices, enabledExtensions);
+    RefPtr<WebCLContext> context = WebCLContext::create(clContextId, this, devices, enabledExtensions);
+    if (!context) {
+        es.throwWebCLException(WebCLException::FAILURE, WebCLException::failureMessage);
+        return nullptr;
+    }
+
+    return context;
+}
+
+void WebCL::waitForEvents(const Vector<RefPtr<WebCLEvent>>& events, WebCLCallback* whenFinished, ExceptionState& es)
+{
+    validateWebCLEventList(events, es, !whenFinished);
+    if (es.hadException()) {
+        es.throwIfNeeded();
+        return;
+    }
+
+    waitForEventsImpl(events, whenFinished);
+}
+
+void WebCL::releaseAll()
+{
+    for (auto webCLContext : m_webCLContexts) {
+        WebCLContext* context = webCLContext.get();
+        if (!context)
+            continue;
+
+        context->releaseAll();
+    }
+}
+
+void WebCL::trackReleaseableWebCLContext(WeakPtr<WebCLContext> context)
+{
+    m_webCLContexts.append(context);
+}
+
+bool WebCL::enableExtension(const String& name)
+{
+    return m_extension.enableExtension(name);
+}
+
+Vector<String> WebCL::getSupportedExtensions()
+{
+    return m_extension.getSupportedExtensions();
+}
+
+void WebCL::getEnabledExtensions(HashSet<String>& extensions)
+{
+    m_extension.getEnabledExtensions(extensions);
+}
+
+WebCL::WebCL()
+{
+    cachePlatforms();
+    cacheSupportedExtensions();
+}
+
+Vector<RefPtr<WebCLCallback>>& WebCL::getCallbacksFromCLEvent(cl_event event)
+{
+    MutexLocker waitForEventLock(m_mutex);
+    m_completedCLEvents.append(event);
+    m_callbacks.clear();
+    for (unsigned j = 0; j < m_callbackRegisterQueue.size(); j++) {
+        Vector<RefPtr<WebCLEvent>> webEvents = m_callbackRegisterQueue[j].first;
+        unsigned i;
+        for (i = 0; i < webEvents.size(); i++) {
+            if (!m_completedCLEvents.contains(webEvents[i]->getEvent()))
+                break;
+        }
+
+        if (i == webEvents.size()) {
+            m_callbacks.append(m_callbackRegisterQueue[j].second);
+            m_callbackRegisterQueue.remove(j--);
+        }
+    }
+
+    return m_callbacks;
+}
+
+void WebCL::callbackProxy(cl_event event, cl_int type, void* userData)
+{
+    WebCL* webcl = static_cast<WebCL*>(userData);
+    if (type != CL_COMPLETE) {
+        return;
+    }
+
+    Vector<RefPtr<WebCLCallback>> callbacks = webcl->getCallbacksFromCLEvent(event);
```

Comment:
It can cause SegFault.
callbacks instance will be deleted in this scope but callOnMainThread() will use it in main thread after deleted. you need to pass this instance as argument.